main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+      *
+      * Returns `Result.Err((ex, l, s))` where `c` is a conflict, `l` is a list of unsolved equations, and `s` is a partial substitution.
+      */
+    def solve(): Result[BoolSubstitution, (FastBoolUnificationException, List[Equation], BoolSubstitution)] = {
+      try {
+        phase0Init()
+        phase1UnitPropagation()
+        phase2VarPropagation()
+        phase3VarAssignment()
+        phase4SVE()
+        // verifySolution() // Not enabled by default due to cost. Enable for debugging.
+        verifySolutionSize()
+
+        Result.Ok(currentSubst)
+      } catch {
+        case ex: ConflictException => Result.Err((ex, currentEqns, currentSubst))
+        case ex: TooComplexException => Result.Err((ex, currentEqns, currentSubst))
+      }
+    }
+
+    private def phase0Init(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Input")
+      debugln("-".repeat(80))
+      printEquations()
+      debugln()
+    }
+
+    private def phase1UnitPropagation(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Phase 1: Unit Propagation")
+      debugln("    (resolves all equations of the form: x = c where x is a var and c is const)")
+      debugln("-".repeat(80))
+      val s = propagateUnit(currentEqns, currentSubst)
+      updateState(s)
+      printEquations()
+      printSubstitution()
+      debugln()
+    }
+
+    private def phase2VarPropagation(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Phase 2: Variable Propagation")
+      debugln("    (resolves all equations of the form: x = y where x and y are vars)")
+      debugln("-".repeat(80))
+      val s = propagateVars(currentEqns, currentSubst)
+      updateState(s)
+      printEquations()
+      printSubstitution()
+      debugln()
+    }
+
+    private def phase3VarAssignment(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Phase 3: Variable Assignment")
+      debugln("    (resolves all equations of the form: x = t where x is free in t)")
+      debugln("-".repeat(80))
+      val s = varAssignment(currentEqns, currentSubst)
+      updateState(s)
+      printEquations()
+      printSubstitution()
+      debugln()
+    }
+
+    private def phase4SVE(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Phase 4: Boolean Unification")
+      debugln("    (resolves all remaining equations using SVE.)")
+      debugln("-".repeat(80))
+      val newSubst = boolUnifyAllPickSmallest(currentEqns)
+      currentEqns = Nil
+      currentSubst = newSubst @@ currentSubst
+      printSubstitution()
+      debugln()
+    }
+
+    /**
+      * Updates the internal state with the given list of pending equations and partial substitution `l`.
+      *
+      * Simplifies the pending equations. Throws [[ConflictException]] if an unsolvable equation is detected.
+      */
+    private def updateState(l: (List[Equation], BoolSubstitution)): Unit = {
+      val (nextEqns, nextSubst) = l
+      currentEqns = simplify(nextEqns)
```

Comment:
Is this simplify call to catch inconsistencies that the phases introduce or is it redundant verification?

Reply:
I have renamed to checkAndSimplify. Cleanest would perhaps be to just simplify.